[Bug]: typescript 5.5.2 cannot find input files with source tsconfig.json

### What happened?

Minimal reproducible example below.

With TS 5.5.2, sources cannot be found:

```
bazel build //...
INFO: Analyzed target //:test (91 packages loaded, 2863 targets configured).
WARNING: /home/tos/.cache/bazel/_bazel_tos/287294dcdaea82c1370bff9bd5064654/external/aspect_rules_ts~~ext~npm_typescript/BUILD.bazel:32:17: input 'package' of @@aspect_rules_ts~~ext~npm_typescript//:.aspect_rules_js/node_modules/typescript is a directory; dependency checking of directories is unsound
ERROR: /home/tos/dh/personal/hack/tsc/BUILD.bazel:3:11: Transpiling TypeScript project @@//:test [tsc -p tsconfig.json] failed: (Exit 2): tsc.sh failed: error executing TsProject command (from target //:test) bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/aspect_rules_ts~~ext~npm_typescript/tsc.sh --skipLibCheck --project tsconfig.json --outDir . --rootDir .

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
error TS18003: No inputs were found in config file '/home/tos/.cache/bazel/_bazel_tos/287294dcdaea82c1370bff9bd5064654/sandbox/linux-sandbox/984/execroot/_main/bazel-out/k8-fastbuild/bin/tsconfig.json'. Specified 'include' paths were '["**/*"]' and 'exclude' paths were '["/home/tos/.cache/bazel/_bazel_tos/287294dcdaea82c1370bff9bd5064654/sandbox/linux-sandbox/984/execroot/_main/bazel-out/k8-fastbuild/bin"]'.
Target //:test failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1.016s, Critical Path: 0.30s
INFO: 4 processes: 3 internal, 1 local.
ERROR: Build did NOT complete successfully
```

When downgrading to 5.4.5:

```
bazel build //...
INFO: Analyzed target //:test (90 packages loaded, 2863 targets configured).
WARNING: /home/tos/.cache/bazel/_bazel_tos/287294dcdaea82c1370bff9bd5064654/external/aspect_rules_ts~~ext~npm_typescript/BUILD.bazel:32:17: input 'package' of @@aspect_rules_ts~~ext~npm_typescript//:.aspect_rules_js/node_modules/typescript is a directory; dependency checking of directories is unsound
INFO: Found 1 target...
Target //:test up-to-date:
  bazel-bin/test.js
INFO: Elapsed time: 0.931s, Critical Path: 0.14s
INFO: 2 processes: 1 internal, 1 local.
INFO: Build completed successfully, 2 total actions
```


### Version

Development (host) and target OS/architectures:

Output of `bazel --version`:

```
bazel --version
bazel 7.2.1
```

Version of the Aspect rules, or other relevant rules from your
`WORKSPACE` or `MODULE.bazel` file:

```bzl
bazel_dep(name = "aspect_rules_ts", version = "2.4.2")

rules_ts_ext = use_extension("@aspect_rules_ts//ts:extensions.bzl", "ext", dev_dependency = True)

rules_ts_ext.deps(
    # TS 5.5.2, broken:
    ts_version="5.5.2",
    ts_integrity="sha512-NcRtPEOsPFFWjobJEtfihkLCZCXZt/os3zf8nTxjVH3RvTSxjrCamJpbExGvYOF+tFHc3pA65qpdwPbzjohhew=="
    # TS 5.4.5, works:
    #ts_version="5.4.5",
    #ts_integrity="sha512-vcI4UpRgg81oIRUFwR0WSIHKt11nJ7SAVlYNIu+QpqeyXP+gpQJy/Z4+F0aGxSE4MqwjyXvW/TzgkLAx2AGHwQ=="
)

use_repo(rules_ts_ext, "npm_typescript")
```

Language(s) and/or frameworks involved:

None

### How to reproduce

MODULE.bazel from above.

BUILD.bazel:

```bzl
load("@aspect_rules_ts//ts:defs.bzl", "ts_project")

ts_project(
    name = "test",
    srcs = ["test.ts"],
)
```

tsconfig.json

```
{}
```

.bazelrc

```
common --@aspect_rules_ts//ts:skipLibCheck=always
common --@aspect_rules_ts//ts:default_to_tsc_transpiler
```

test.ts

```
console.log("foo");
```

Then run `bazel build //:test`


### Any other information?

The `tsconfig.json` must be a source file.

If you do (and optionally delete `tsconfig.json`), the issue vanishes.

```diff
 ts_project(
     name = "test",
     srcs = ["test.ts"],
+    tsconfig = {},
 )
```

Specifying `exclude: []` fixes the issue. It looks like typescript 5.5 adds the `outDir` passed via command line to `exclude`.

typescript 5.4 doesn't do that as can be shown experimentally (by setting a bogus include pattern):

TS 5.4:

```
error TS18003: No inputs were found in config file '/home/tos/.cache/bazel/_bazel_tos/287294dcdaea82c1370bff9bd5064654/sandbox/linux-sandbox/1051/execroot/_main/bazel-out/k8-fastbuild/bin/src/tsconfig.json'. Specified 'include' paths were '["**/*.tsx"]' and 'exclude' paths were '[]'.
```

TS 5.5:

```
error TS18003: No inputs were found in config file '/home/tos/.cache/bazel/_bazel_tos/287294dcdaea82c1370bff9bd5064654/sandbox/linux-sandbox/1054/execroot/_main/bazel-out/k8-fastbuild/bin/src/tsconfig.json'. Specified 'include' paths were '["**/*.tsx"]' and 'exclude' paths were '["/home/tos/.cache/bazel/_bazel_tos/287294dcdaea82c1370bff9bd5064654/sandbox/linux-sandbox/1054/execroot/_main/bazel-out/k8-fastbuild/bin/src"]'.
```

So a workaround is to override exclude manually.

--- a/ts/defs.bzl
+++ b/ts/defs.bzl
@@ -298,6 +298,13 @@ def ts_project(
         declaration_dir = compiler_options.pop("declarationDir", declaration_dir)
         root_dir = compiler_options.pop("rootDir", root_dir)
 
+        # When generating a tsconfig.json and we set rootDir we need to add the exclude field,
+        # because of these tickets (validation for not generated tsconfig is also present elsewhere):
+        # https://github.com/microsoft/TypeScript/issues/59036 and
+        # 'https://github.com/aspect-build/rules_ts/issues/644
+        if root_dir != None and "exclude" not in tsconfig:
+            tsconfig["exclude"] = []
+
         if srcs == None:
             # Default sources based on macro attributes after applying tsconfig properties
             srcs = _default_srcs(

--- a/ts/private/ts_lib.bzl
+++ b/ts/private/ts_lib.bzl
@@ -227,6 +227,8 @@ def _to_out_path(f, out_dir, root_dir):
 
 def _to_js_out_paths(srcs, out_dir, root_dir, allow_js, resolve_json_module, ext_map, default_ext):
     outs = []
+    out_dir = _remove_leading_dot_slash(out_dir)
+    root_dir = _remove_leading_dot_slash(root_dir)
     for f in srcs:
         if _is_ts_src(f, allow_js, resolve_json_module, False):
             out = _to_out_path(f, out_dir, root_dir)
@@ -324,6 +326,9 @@ def _declare_outputs(ctx, paths):
         ctx.actions.declare_file(path)
         for path in paths
     ]
+
+def _remove_leading_dot_slash(string_path):
+    return string_path.removeprefix("./") if string_path else string_path
 
 lib = struct(
     declare_outputs = _declare_outputs,

--- a/ts/private/ts_project_options_validator.js
+++ b/ts/private/ts_project_options_validator.js
@@ -113,6 +113,26 @@ function main(_a) {
             )
         }
     }
+    function checkRootDirExclude() {
+        var rootDirAttrValue = attrs["root_dir"];
+        var outDirAttrValue = attrs["out_dir"];
+        var excludeOptionValue = config["exclude"];
+
+        let rootDirNotEmpty = rootDirAttrValue !== undefined && rootDirAttrValue !== "";
+        let outDirEmpty = outDirAttrValue === undefined || outDirAttrValue === "";
+
+        if (rootDirNotEmpty && outDirEmpty && excludeOptionValue === undefined) {
+            throw new Error(
+                '\n\nWhen root dir is set, exclude must also be set to empty array in the tsconfig file.\n\n' +
+                'For example, tsconfig.json:\n' +
+                '{\n' +
+                '   "exclude": []\n' +
+                '}\n\n' +
+                'See tickets: https://github.com/microsoft/TypeScript/issues/59036 and ' +
+                'https://github.com/aspect-build/rules_ts/issues/644\n'
+            )
+        }
+    }
     var jsxEmit =
         ((_b = {}),
         (_b[ts.JsxEmit.None] = 'none'),
@@ -175,6 +195,7 @@ function main(_a) {
     check('incremental')
     check('tsBuildInfoFile', 'ts_build_info_file')
     check_out_dir()
+    checkRootDirExclude()
     check_nocheck()
     check_preserve_jsx()
     if (failures.length > 0) {

--- a/ts/private/ts_validate_options.bzl
+++ b/ts/private/ts_validate_options.bzl
@@ -26,6 +26,7 @@ Assumes all tsconfig file deps are already copied to the bin directory.
         incremental = ctx.attr.incremental,
         ts_build_info_file = ctx.attr.ts_build_info_file,
         isolated_typecheck = ctx.attr.isolated_typecheck,
+        root_dir = ctx.attr.root_dir,
     )
     arguments.add_all([
         to_output_relative_path(tsconfig),

--- a/ts/test/ts_project_test.bzl
+++ b/ts/test/ts_project_test.bzl
@@ -170,12 +170,32 @@ def ts_project_test_suite(name):
         targets = [":wrapper.d.ts.map"],
     )
 
+    write_file(
+        name = "dirty_out_dir_ts",
+        out = "dirty_out_dir.ts",
+        content = ["console.log(1)"],
+        tags = ["manual"],
+    )
+    ts_project_wrapper(
+        name = "dirty_out_dir",
+        srcs = ["dirty_out_dir.ts"],
+        out_dir = "./out-dir",
+        tsconfig = _TSCONFIG,
+        tags = ["manual"],
+    )
+
+    build_test(
+        name = "dirty_out_dir_test",
+        targets = [":dirty_out_dir"],
+    )
+
     native.test_suite(
         name = name,
         tests = [
             ":dir_test",
             ":use_dir_test",
             ":wrapper_test",
+            ":dirty_out_dir_test",
         ],
     )
 

